### DISCUSSION TITLE
Ping/pong implementation with Apollo Server Express

### DISCUSSION THREAD

**Discussion by BeauGieskens** [enisdenjo/graphql-ws#278] [Q&A] [answered]
We noticed that setting up the subprotocol ping/pong example is just for when using this library on its own. We've found it very difficult to try and figure it out ourselves. We started with the example code from [here](https://github.com/enisdenjo/graphql-ws#apollo-server-express).

We're able to follow the example up until it's deriving logic and message types from the graphql-ws server, at which point the Express backend doesn't seem to give us the control needed to direct messages appropriately.

Is it even possible to set up the ping/pong messages with an Apollo Server Express configuration?

**Comment by enisdenjo** [comment, marked as answer]
Please read https://github.com/enisdenjo/graphql-ws/discussions/231#discussioncomment-1301945 and https://github.com/enisdenjo/graphql-ws/discussions/231#discussioncomment-1305810 in detail first (and optionally about why the client doesn't need to ping the server: https://github.com/enisdenjo/graphql-ws/issues/117#issuecomment-784352250), before proceeding with subprotocol pings/pongs.

TL;DR you most probably **do not need them**.

> Is it even possible to set up the ping/pong messages with an Apollo Server Express configuration?

Yes. You have to make your own `graphql-ws` WebSocket server as showcased in the ["server usage with ws and subprotocol pings and pongs" recipe](https://github.com/enisdenjo/graphql-ws#ws-sub-ping-pong) and use it instead of the `useServer` in your Apollo Server.
